### PR TITLE
limit the number of "remembered" files

### PR DESCRIPTION
**Pull request by leorosa**
With this PR, the number of file names/positions stored is limited.
Otherwise, the log file will grow indefinitely, storing all files, even the ones that do not exist anymore.

### CODE REVIEW

**Comment by erf** [comment]
Is there actually an performance issue regarding reading the cursors file ?
Do you really need the `files` map and why start at `n = 2` ?
Why remove the logic for sorting paths and  concatenating path strings ?
Not sure about rereading the whole file each time you set a cursors .. also you are increasing the max size limit there? I agree some other instance could open a vis file, but not sure how common it is with multiple vis open or if we should care, if so it should be a seperate commit.

If i wanted to do a change like this, i would make the `maxfiles` limit configurable by adding it to the module `M`.
I would also perhaps keep a date per cursor, and sort them so the least used would be re-used. Or just make a simpler change by skipping an existing loop by a max size.

I think there is too many changes at once here, and not sure i agree with / understand all of them.

This made me think of another solution though to limit the file size. I see there are many repeated folder names that i think could be optimized away to save disk space and make faster read / write.

**Comment by erf** [comment]
I made a new commit which compresses the `.vis-cursors` file now by replacing repeating folders.

Will close this for now ref my comments.

**Comment by leorosa** [comment]
Sure, I should have explained the changes made.

> Is there actually an performance issue regarding reading the cursors file ?

I have not noticed any performance issue on my (10 years old) notebook.

> Do you really need the files map and why start at n = 2 ?

I think that the `files` map is needed, because (I read it somewhere that) there is no guarantee that the keys in the `cursors` map can be retrieved in order. `files[1]` refers to the file being edited, and the ones present in the log file are inserted beginning in key 2.

> Why remove the logic for sorting paths and concatenating path strings ?

Because... this way, I am considering that the last edited files are at the top of the list, without the need to save/compare dates. Moreover, while I was not noticing performance issues, I believe that a large list of files may be slower to `sort`

> Not sure about rereading the whole file each time you set a cursors .. also you are increasing the max size limit there?

Yes, if the file currently being edited is already present in the list, I'm disregarding it (`files[key] = nil`, in which case this entry will be ignored in `write_cursors()`), and enabling another one to be also stored (hence increasing `maxfiles`).

> I agree some other instance could open a vis file, but not sure how common it is with multiple vis open or if we should care, if so it should be a seperate commit.

Sometimes I have more than one vis instance running. I thought that it would be more correct to have the log file always up to date.

> If i wanted to do a change like this, i would make the maxfiles limit configurable by adding it to the module M.

I have not thought about that. It is a nice idea - perhaps setting it to `0` to not impose any limit...

**Comment by erf** [comment]
I've added reading the cursors file before closing a window in the latest commit as you suggested.

The other changes i think will be difficult to implement now i think, as we need to keep the list sorted in order to compress the folders. 

BTW my point regarding performance is that i'm not sure there is a need to limit the number of files / lines we write / read, as that number could become quite large without problems; we could perhaps test this

**Comment by erf** [comment]
I'll take a second look at this ..

File: init.lua [hunk 1]
```diff
@@ -33,48 +35,60 @@ end
...
 	if f == nil then
 		return
 	end
 	-- read positions per file path
+	local n = 2 -- [n] is reserved to the current file
 	for line in f:lines() do
 		for path, pos in string.gmatch(line, '(.+)[,%s](%d+)') do
 			cursors[path] = pos
+			files[n] = path
+			n = n+1
```
Comment (erf) [inline init.lua:49]: Instead of doing this we could do table.insert(files, path)

File: init.lua [hunk 2]
```diff
@@ -33,48 +35,60 @@ end
...
-	table.sort(paths)
-	-- buffer cursors string
-	local t = {}
-	for i, path in ipairs(paths) do
-		table.insert(t, string.format('%s,%d', path, cursors[path]))
+	for n=1,M.maxcursors do
+		if files[n] then
+			f:write(string.format('%s,%d\n', files[n], cursors[files[n]]))
+		end
 	end
```
Comment (erf) [inline init.lua:65]: We could still buffer string here

File: init.lua [hunk 3]
```diff
@@ -33,48 +35,60 @@ end
...
-	local s = table.concat(t, '\n')
-	f:write(s)
 	f:close()
 end
 
 local set_cursor_pos = function(win)
+	read_cursors() -- reread M.path, which may be changed meanwhile by another vis instance
```
Comment (erf) [inline init.lua:70]: This should be a separate commit

File: init.lua [hunk 4]
```diff
@@ -33,48 +35,60 @@ end
...
 	if win.file == nil or win.file.path == nil then
 		return
 	end
 	if not file_exists(win.file.path) then
 		return
 	end
+	for key,val in pairs(files) do
```
Comment (erf) [inline init.lua:77]: Not sure what this is for

File: init.lua [hunk 5]
```diff
@@ -33,48 +35,60 @@ end
...
+		if val == win.file.path then
+			files[key] = nil
+			M.maxcursors = M.maxcursors+1
+		end
+	end
+	files[1] = win.file.path
```
Comment (erf) [inline init.lua:83]: Why do we need to store the current file here?

**Comment by erf** [comment]
@leorosa I had a second look at your PR and i think this could be a nice addition, but i think it could be simplified somewhat. You could use a `table.insert` instead of the n variable. I don't think you need to keep the 1 index a special case, just check after reading it, if it exists from before and then add it to the top using `table.insert` and remove the old item. Then you don't need the check in `set_cursor_pos`. 

You'd need to rebase to the latest in order for this change, and please add the `read_cursors` call inside `set_cursor_pos` as a separate commit.

**Comment by erf** [comment]
@leorosa I've solved the "max number of files" now in commit 2bb1152 and 0e2808e using my own implementation inspired by yours. Please give it a try if you want. Thanks for your help
